### PR TITLE
Implementation of Semi-Supervised Learning with 1 training set instead of 2 different training subsets

### PR DESCRIPTION
**Pull request by lucamarini22**
## Main commits

In this PR I want to implement SSL by using a single training set (the full training set $D$). In particular, for each patch:

- Its labeled pixels will contribute to the supervised loss.
- Its unlabeled pixels will contributed to the unsupervised loss.

The SSL mode that is already implemented instead uses 2 training subsets $D_s$ and $D_u$, s.t. $D_s \ D_u = D$, where $D$ is the full training set. In this implementation:

- Patches in $D_s$ are used to compute the supervised loss.
- Patches in $D_u$ are used to compute the unsupervised loss.

## Additional commits
- Update Readme, docs, fix bugs, improve readability

### CODE REVIEW

File: inference.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    checkpoint = torch.load(model_file, map_location=device)
+    model.load_state_dict(checkpoint)
+
+    del checkpoint  # dereference
+    if torch.cuda.is_available():
+        torch.cuda.empty_cache()
+    
+    model.eval()
+
+    image = ...
```
[the remark this answers is not in the document]
Comment (lucamarini22) [reply inference.py:81]: I delete this script later

**Review by lucamarini22 — COMMENTED** [review]
ok
